examples/window-app/nrfconnect/main/WindowCovering.cpp

```diff
@@ -0,0 +1,340 @@
+    {
+        chip::DeviceLayer::PlatformMgr().ScheduleWork(DriveCurrentTiltPosition);
+    }
+}
+
+void WindowCovering::DriveCurrentLiftPosition(intptr_t)
+{
+    NPercent100ths current{};
+    NPercent100ths target{};
+    NPercent100ths positionToSet{};
+
+    VerifyOrReturn(Attributes::CurrentPositionLiftPercent100ths::Get(Endpoint(), current) == EMBER_ZCL_STATUS_SUCCESS);
+    VerifyOrReturn(Attributes::TargetPositionLiftPercent100ths::Get(Endpoint(), target) == EMBER_ZCL_STATUS_SUCCESS);
+
+    UpdateOperationalStatus(MoveType::LIFT, ComputeOperationalState(target, current));
+
+    positionToSet.SetNonNull(CalculateSingleStep(MoveType::LIFT));
+    LiftPositionSet(Endpoint(), positionToSet);
+
+    // assume single move completed
+    Instance().mInLiftMove = false;
+
+    VerifyOrReturn(Attributes::CurrentPositionLiftPercent100ths::Get(Endpoint(), current) == EMBER_ZCL_STATUS_SUCCESS);
+
+    if (!TargetCompleted(MoveType::LIFT, current, target))
+    {
+        // continue to move
+        StartTimer(MoveType::LIFT, sMoveTimeoutMs);
+    }
+    else
+    {
+        // the OperationalStatus should indicate no-lift-movement after the target is completed
+        UpdateOperationalStatus(MoveType::LIFT, ComputeOperationalState(target, current));
+    }
+}
+
+chip::Percent100ths WindowCovering::CalculateSingleStep(MoveType aMoveType)
+{
+    EmberAfStatus status{};
+    chip::Percent100ths percent100ths{};
+    NPercent100ths current{};
+    OperationalState opState{};
+
+    OperationalStatus opStatus = OperationalStatusGet(Endpoint());
+
+    if (aMoveType == MoveType::LIFT)
+    {
+        status  = Attributes::CurrentPositionLiftPercent100ths::Get(Endpoint(), current);
+        opState = opStatus.lift;
+    }
+    else if (aMoveType == MoveType::TILT)
+    {
+        status  = Attributes::CurrentPositionTiltPercent100ths::Get(Endpoint(), current);
+        opState = opStatus.tilt;
+    }
+
+    if ((status == EMBER_ZCL_STATUS_SUCCESS) && !current.IsNull())
+    {
+        static constexpr auto sPercentDelta{ WC_PERCENT100THS_MAX_CLOSED / 20 };
+        percent100ths = ComputePercent100thsStep(opState, current.Value(), sPercentDelta);
+    }
+    else
+    {
+        LOG_ERR("Cannot read the current lift position. Error: %d", static_cast<uint8_t>(status));
+    }
+
+    return percent100ths;
+}
+
+bool WindowCovering::TargetCompleted(MoveType aMoveType, NPercent100ths aCurrent, NPercent100ths aTarget)
+{
+    OperationalStatus currentOpStatus = OperationalStatusGet(Endpoint());
+    OperationalState currentOpState   = (aMoveType == MoveType::LIFT) ? currentOpStatus.lift : currentOpStatus.tilt;
+
+    if (!aCurrent.IsNull() && !aTarget.IsNull())
+    {
+        switch (currentOpState)
+        {
+        case OperationalState::MovingDownOrClose:
+            return (aCurrent.Value() >= aTarget.Value());
+        case OperationalState::MovingUpOrOpen:
+            return (aCurrent.Value() <= aTarget.Value());
```

Comment:
Why not reuse ComputeOperationalState here and look at the OperationalState::Stall 
or a diff with currentOpState != ComputeOperationalState( to avoid this switch

Reply:
I hope we can look at this on the occasion of other refactoring actions.